### ISSUE TITLE
Support Python 3.13

### ISSUE THREAD

**Issue by andrey-khropov** [catboost/catboost#2748]
Python 3.13 is expected to be released soon (2024-10-02): https://peps.python.org/pep-0719/

**Comment by bvdmitri** [comment]
Cannot install `catboost` on MacOS with fresh python 3.13.0

```
Using cached catboost-1.2.7.tar.gz (71.5 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [24 lines of output]
      Traceback (most recent call last):
        File "/Users/bvdmitri/Projects/LazyDynamics/Litigation/litigation/.venv/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
          ~~~~^^
        File "/Users/bvdmitri/Projects/LazyDynamics/Litigation/litigation/.venv/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/bvdmitri/Projects/LazyDynamics/Litigation/litigation/.venv/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/private/var/folders/x3/rqw10ky529915v7_l4t7vz0m0000gn/T/pip-build-env-cn06qfv9/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 334, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/x3/rqw10ky529915v7_l4t7vz0m0000gn/T/pip-build-env-cn06qfv9/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 304, in _get_build_requires
          self.run_setup()
          ~~~~~~~~~~~~~~^^
        File "/private/var/folders/x3/rqw10ky529915v7_l4t7vz0m0000gn/T/pip-build-env-cn06qfv9/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 522, in run_setup
          super().run_setup(setup_script=setup_script)
          ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/x3/rqw10ky529915v7_l4t7vz0m0000gn/T/pip-build-env-cn06qfv9/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 320, in run_setup
          exec(code, locals())
          ~~~~^^^^^^^^^^^^^^^^
        File "<string>", line 733, in <module>
        File "<string>", line 205, in get_catboost_version
      KeyError: 'VERSION'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by Carlos-Ot** [comment]
I have the same problem installing it on Windows with the latest Python version 3.13.0

```
Collecting catboost
  Using cached catboost-1.2.7.tar.gz (71.5 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [24 lines of output]
      Traceback (most recent call last):
        File "C:\Users\cotto\AppData\Local\Programs\Python\Python313\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
          ~~~~^^
        File "C:\Users\cotto\AppData\Local\Programs\Python\Python313\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\cotto\AppData\Local\Programs\Python\Python313\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "C:\Users\cotto\AppData\Local\Temp\pip-build-env-pw1w7_lc\overlay\Lib\site-packages\setuptools\build_meta.py", line 334, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\cotto\AppData\Local\Temp\pip-build-env-pw1w7_lc\overlay\Lib\site-packages\setuptools\build_meta.py", line 304, in _get_build_requires
          self.run_setup()
          ~~~~~~~~~~~~~~^^
        File "C:\Users\cotto\AppData\Local\Temp\pip-build-env-pw1w7_lc\overlay\Lib\site-packages\setuptools\build_meta.py", line 522, in run_setup
          super().run_setup(setup_script=setup_script)
          ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\cotto\AppData\Local\Temp\pip-build-env-pw1w7_lc\overlay\Lib\site-packages\setuptools\build_meta.py", line 320, in run_setup
          exec(code, locals())
          ~~~~^^^^^^^^^^^^^^^^
        File "<string>", line 733, in <module>
        File "<string>", line 205, in get_catboost_version
      KeyError: 'VERSION'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.
```

**Comment by andrey-khropov** [comment]
> I have the same problem installing it on Windows with the latest Python version 3.13.0
> 
> ```
> Collecting catboost
>   Using cached catboost-1.2.7.tar.gz (71.5 MB)
>   Installing build dependencies ... done
>   Getting requirements to build wheel ... error
>   error: subprocess-exited-with-error
> 
>   × Getting requirements to build wheel did not run successfully.
>   │ exit code: 1
>   ╰─> [24 lines of output]
>       Traceback (most recent call last):
>         File "C:\Users\cotto\AppData\Local\Programs\Python\Python313\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
>           main()
>           ~~~~^^
>         File "C:\Users\cotto\AppData\Local\Programs\Python\Python313\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
>           json_out['return_val'] = hook(**hook_input['kwargs'])
>                                    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
>         File "C:\Users\cotto\AppData\Local\Programs\Python\Python313\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 118, in get_requires_for_build_wheel
>           return hook(config_settings)
>         File "C:\Users\cotto\AppData\Local\Temp\pip-build-env-pw1w7_lc\overlay\Lib\site-packages\setuptools\build_meta.py", line 334, in get_requires_for_build_wheel
>           return self._get_build_requires(config_settings, requirements=[])
>                  ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "C:\Users\cotto\AppData\Local\Temp\pip-build-env-pw1w7_lc\overlay\Lib\site-packages\setuptools\build_meta.py", line 304, in _get_build_requires
>           self.run_setup()
>           ~~~~~~~~~~~~~~^^
>         File "C:\Users\cotto\AppData\Local\Temp\pip-build-env-pw1w7_lc\overlay\Lib\site-packages\setuptools\build_meta.py", line 522, in run_setup
>           super().run_setup(setup_script=setup_script)
>           ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "C:\Users\cotto\AppData\Local\Temp\pip-build-env-pw1w7_lc\overlay\Lib\site-packages\setuptools\build_meta.py", line 320, in run_setup
>           exec(code, locals())
>           ~~~~^^^^^^^^^^^^^^^^
>         File "<string>", line 733, in <module>
>         File "<string>", line 205, in get_catboost_version
>       KeyError: 'VERSION'
>       [end of output]
> 
>   note: This error originates from a subprocess, and is likely not a problem with pip.
> error: subprocess-exited-with-error
> 
> × Getting requirements to build wheel did not run successfully.
> │ exit code: 1
> ╰─> See above for output.
> ```

This is because of this breaking change: https://github.com/python/cpython/issues/118888 that affected code used in `setup.py`. I'll fix this.

**Comment by G-SaiVishwas** [comment]
Same problem here on latest version of python on MacOs
```
Collecting catboost
  Using cached catboost-1.2.7.tar.gz (71.5 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [24 lines of output]
      Traceback (most recent call last):
        File "/Users/saivishwasgooty/.local/pipx/venvs/pip/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
          ~~~~^^
        File "/Users/saivishwasgooty/.local/pipx/venvs/pip/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/saivishwasgooty/.local/pipx/venvs/pip/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/private/var/folders/20/g663rsm54hnfnq9q1vbz72zw0000gn/T/pip-build-env-9yxsspwe/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 334, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/20/g663rsm54hnfnq9q1vbz72zw0000gn/T/pip-build-env-9yxsspwe/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 304, in _get_build_requires
          self.run_setup()
          ~~~~~~~~~~~~~~^^
        File "/private/var/folders/20/g663rsm54hnfnq9q1vbz72zw0000gn/T/pip-build-env-9yxsspwe/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 522, in run_setup
          super().run_setup(setup_script=setup_script)
          ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/20/g663rsm54hnfnq9q1vbz72zw0000gn/T/pip-build-env-9yxsspwe/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 320, in run_setup
          exec(code, locals())
          ~~~~^^^^^^^^^^^^^^^^
        File "<string>", line 733, in <module>
        File "<string>", line 205, in get_catboost_version
      KeyError: 'VERSION'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by calvelli** [comment]
I am facing the exact sample problem on windows 10 Pro, python 3.13.1 while trying to install catboost 1.2.7. Any advancements/fixes or workaround?

**Comment by andrey-khropov** [comment]
> I am facing the exact sample problem on windows 10 Pro, python 3.13.1 while trying to install catboost 1.2.7. Any advancements/fixes or workaround?


This particular problem has been fixed by https://github.com/catboost/catboost/pull/2803 but it does not fully fix this issue because:

- Numpy officially supports python 3.13 only [since release 2.1.0](https://numpy.org/news/#numpy-210-released). It means this issue depends on #2671 
- CatBoost uses Cython and currently only Cython 0.29.x is supported. But Cython 3.x is needed for code generation for Python 3.13. I've created a separate issue for this: #2810

**Comment by thiagobdp** [comment]
Same on Windows 11 Pro, Python 3.13.2

```
Collecting catboost==1.2.7
  Using cached catboost-1.2.7.tar.gz (71.5 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [24 lines of output]
      Traceback (most recent call last):
        File "C:\Users\W507417\PycharmProjects\ps-ag-fraud-ai\.venv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 389, in <module>
          main()
          ~~~~^^
        File "C:\Users\W507417\PycharmProjects\ps-ag-fraud-ai\.venv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 373, in main
          json_out["return_val"] = hook(**hook_input["kwargs"])
                                   ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\W507417\PycharmProjects\ps-ag-fraud-ai\.venv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 143, in get_requires_for_build_wheel
          return hook(config_settings)
        File "C:\Users\W507417\AppData\Local\Temp\pip-build-env-rwzk8zfz\overlay\Lib\site-packages\setuptools\build_meta.py", line 334, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\W507417\AppData\Local\Temp\pip-build-env-rwzk8zfz\overlay\Lib\site-packages\setuptools\build_meta.py", line 304, in _get_build_requires
          self.run_setup()
          ~~~~~~~~~~~~~~^^
        File "C:\Users\W507417\AppData\Local\Temp\pip-build-env-rwzk8zfz\overlay\Lib\site-packages\setuptools\build_meta.py", line 522, in run_setup
          super().run_setup(setup_script=setup_script)
          ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\W507417\AppData\Local\Temp\pip-build-env-rwzk8zfz\overlay\Lib\site-packages\setuptools\build_meta.py", line 320, in run_setup
          exec(code, locals())
          ~~~~^^^^^^^^^^^^^^^^
        File "<string>", line 733, in <module>
        File "<string>", line 205, in get_catboost_version
      KeyError: 'VERSION'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by andrey-khropov** [comment]
> Is there any update on this issue?

Python 3.13 is supported in the master branch since https://github.com/catboost/catboost/commit/b6fd136e3fc173bc335ae65fb976f27b550f24ad . The release will be very soon.

**Comment by andrey-khropov** [comment]
Note that right now a variant of CPython 3.13 with free threading is not supported. See ##2856 about free threading.
